src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs

```diff
@@ -2,50 +2,72 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
 using System;
 using System.Collections.Immutable;
 using System.Diagnostics;
 using System.Linq;
 using System.Reflection.Metadata;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
-using Microsoft.CodeAnalysis.PooledObjects;
 
 namespace Microsoft.CodeAnalysis.CSharp.CodeGen
 {
     internal partial class CodeGenerator
     {
-        private void EmitStackAllocInitializers(TypeSymbol type, BoundArrayInitialization inits)
+        private void EmitStackAlloc(TypeSymbol type, BoundArrayInitialization? inits, BoundExpression count)
         {
+            if (inits is null)
+            {
+                emitLocalloc();
+                return;
+            }
+
             Debug.Assert(type is PointerTypeSymbol || type is NamedTypeSymbol);
+            Debug.Assert(_diagnostics.DiagnosticBag is not null);
 
             var elementType = (type.TypeKind == TypeKind.Pointer
                 ? ((PointerTypeSymbol)type).PointedAtTypeWithAnnotations
                 : ((NamedTypeSymbol)type).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0]).Type;
 
+            bool isReadOnlySpan = TypeSymbol.Equals(
+                (type as NamedTypeSymbol)?.OriginalDefinition, _module.Compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T), TypeCompareKind.ConsiderEverything);
+
             var initExprs = inits.Initializers;
 
-            var initializationStyle = ShouldEmitBlockInitializerForStackAlloc(elementType, initExprs);
-            if (initializationStyle == ArrayInitializerStyle.Element)
+            bool isEncDelta = _module.IsEncDelta;
+            var initializationStyle = ShouldEmitBlockInitializerForStackAlloc(elementType, initExprs, isEncDelta);
+
+            if (isReadOnlySpan)
             {
+                var createSpanHelper = getCreateSpanHelper(_module, elementType);
+
+                // ROS<T> is only used here if it has already been decided to use CreateSpan
+                Debug.Assert(createSpanHelper is not null);
+                Debug.Assert(UseCreateSpanForReadOnlySpanStackAlloc(elementType, inits, isEncDelta: isEncDelta));
+
+                EmitExpression(count, used: false);
+
+                ImmutableArray<byte> data = GetRawData(initExprs);
+                _builder.EmitCreateSpan(data, createSpanHelper, inits.Syntax, _diagnostics.DiagnosticBag);
```

Comment:
>_builder.EmitCreateSpan(data, createSpanHelper, inits.Syntax, _diagnostics.DiagnosticBag); [](http://example.com/codeflow?start=16&length=90)

It looks like the issue that we are trying to address is specifically about optimizing initialization of the block allocated on the stack. It looks like this code path is doing something different completely.

Reply:
A lot of different things can be discussed and prototyped, this doesn't mean that they should be implemented in the compiler. This part of the change is out of scope of the the issue being addressed and it is in direct violation of the language specification. I think it should simply be reverted (not simply moved to a different PR), because it is not merely an optimization (has an observable difference in behavior). If we want to change semantics of a stack allocation expression, that should go through the regular process of a language change: proposal on [https://github.com/dotnet/csharplang](https://github.com/dotnet/csharplang), etc.